kafka-rest/src/test/java/io/confluent/kafkarest/integration/AbstractConsumerTest.java

```diff
@@ -206,33 +220,46 @@ protected String startConsumeMessages(
       GenericType<List<RecordType>> responseEntityType,
       Converter converter,
       Function<RecordType, ConsumerRecord<ClientK, ClientV>> fromJsonWrapper) {
-    Response response = request(instanceUri + "/records").accept(accept).get();
-    assertOKResponse(response, responseMediatype);
-    List<RecordType> consumed = TestUtils.tryReadEntityOrLog(response, responseEntityType);
-    try {
-      assertEquals(records.size(), consumed.size());
-    } catch (AssertionError e) {
-      pause();
-      Response response2 = request(instanceUri + "/records").accept(accept).get();
-      assertOKResponse(response2, responseMediatype);
-      List<RecordType> consumed2 = TestUtils.tryReadEntityOrLog(response2, responseEntityType);
-      consumed.addAll(consumed2);
+
+    int retryCount = 0;
+    final int TOTAL_RETRIES = 3;
+    List<RecordType> consumed = new ArrayList<>();
+
+    while (retryCount < TOTAL_RETRIES) {
+      Response response = request(instanceUri + "/records").accept(accept).get();
+      assertOKResponse(response, responseMediatype);
+      consumed.addAll(TestUtils.tryReadEntityOrLog(response, responseEntityType));
       try {
         assertEquals(records.size(), consumed.size());
-      } catch (AssertionError e2) {
-        pause(2);
-        Response response3 = request(instanceUri + "/records").accept(accept).get();
-        assertOKResponse(response3, responseMediatype);
-        List<RecordType> consumed3 = TestUtils.tryReadEntityOrLog(response3, responseEntityType);
-        consumed.addAll(consumed3);
-        assertEquals(records.size(), consumed.size());
+        retryCount = TOTAL_RETRIES;
+      } catch (AssertionError assertionError) {
+        logConsumeData(records, consumed);
+        pause(++retryCount);
       }
     }
 
     assertEqualsMessages(
         records, consumed.stream().map(fromJsonWrapper).collect(Collectors.toList()), converter);
   }
 
+  private final <KafkaK, KafkaV, RecordType> void logConsumeData(
+      List<ProducerRecord<KafkaK, KafkaV>> records, List<RecordType> consumed) {
+    log.info("Records produced : {}", records.size());
+    log.info("Consume failed.  Record offsets are:");
+    consumed.forEach(
+        record -> {
+          if (record instanceof SchemaConsumerRecord) {
+            log.info("{}", ((SchemaConsumerRecord) record).getOffset());
+          } else if (record instanceof BinaryConsumerRecord) {
+            log.info("{}", ((BinaryConsumerRecord) record).getOffset());
+          } else if (record instanceof JsonConsumerRecord) {
+            log.info("{}", ((JsonConsumerRecord) record).getOffset());
```

Comment:
nit: Does it make sense to distinguish between the different record types? It shouldn't cost us anything, should it?

Reply:
The different type of records don't extend/implement any super class. Ie there is no ConsumerRecord common parent that we can cast the unknown-typed record to to be able to call getOffset on it.  And without the cast, all you can call on record is Object's methods :(

So we have to deal with them one at a time.